cellarium/ml/models/geneformer.py

```diff
@@ -117,6 +117,42 @@ def tokenize(self, x_ng: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
         input_ids[~attention_mask] = 0
         return input_ids, attention_mask
 
+    def tokenize_with_perturbations(
+        self,
+        x_ng: torch.Tensor,
+        feature_deletion: list[str] | None = None,
+        feature_activation: list[str] | None = None,
+        feature_map: dict[str, int] | None = None,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+
+        # activation and deletion happen before sorting
+        if feature_deletion:
+            assert all(
+                [g in self.var_names_g for g in feature_deletion]
+            ), "Some feature_deletion elements are not in self.var_names_g"
+            deletion_logic_g = np.logical_or.reduce([(self.var_names_g == g) for g in feature_deletion])
+            x_ng[:, deletion_logic_g] = 0
+        if feature_activation:
+            max_val = x_ng.max()
+            for i, g in enumerate(feature_activation[::-1]):
+                feature_logic_g = self.var_names_g == g
+                assert feature_logic_g.sum() == 1, f"feature_activation element {g} is not in self.var_names_g"
+                top_rank_value = max_val + i + 1
+                x_ng[:, feature_logic_g] = top_rank_value
+
+        # tokenize and sort to give rank-ordered inputs
+        input_ids, attention_mask = self.tokenize(x_ng)
+
+        # feature map is applied after tokenization and sorting
+        if feature_map:
+            for g, target_token in feature_map.items():
+                feature_logic_g = self.var_names_g == g
+                assert feature_logic_g.sum() == 1, f"feature_map key {g} not in self.var_names_g"
+                initial_token = self.feature_ids[feature_logic_g]
```

Comment:
I wonder if there would be any benefits of having a `gene_name -> gene_id` dictionary so we don't have to do this